include/sacn/cpp/universe_discovery.h

```diff
@@ -0,0 +1,399 @@
+ * }
+ * @endcode
+ */
+
+class UniverseDiscovery
+{
+public:
+  /** A handle type used by the sACN library to identify Universe Discovery listener instances. */
+  using Handle = sacn_universe_discovery_t;
+  /** An invalid Handle value. */
+  static constexpr Handle kInvalidHandle = SACN_UNIVERSE_DISCOVERY_INVALID;
+
+  /**
+   * @ingroup sacn_universe_discovery_cpp
+   * @brief A base class for a class that receives notification callbacks from a sACN Universe Discovery listener.
+   */
+  class NotifyHandler
+  {
+  public:
+    virtual ~NotifyHandler() = default;
+
+    /**
+     * @brief Notify that a source is new or has changed.
+     *
+     * This passes the source's current universe list, but you will only get this callback when the module detects
+     * that the source is new or the list has somehow changed.
+     *
+     * The list of sourced universes is guaranteed by the protocol to be numerically sorted.
+     *
+     * @param[in] cid The CID of the source.
+     * @param[in] name The UTF-8 name string.
+     * @param[in] sourced_universes Numerically sorted array of the currently sourced universes.  Will be NULL if the
+     * source is not currently transmitting any universes.
+     * @param[in] num_sourced_universe Size of the sourced_universes array.  Will be 0 if the source is not currently
+     * transmitting any universes.
+     */
+    virtual void HandleUpdateSource(const etcpal::Uuid& cid, const std::string& name, const uint16_t* sourced_universes,
+                                    size_t num_sourced_sources) = 0;
+
+    /**
+     * @brief Notify that a source is no longer transmitting Universe Discovery messages.
+     *
+     * @param[in] cid The CID of the source.
+     * @param[in] name The UTF-8 name string.
+     */
+    virtual void HandleSourceExpired(const etcpal::Uuid& cid, const std::string& name) = 0;
+
+    /**
+     * @brief Notify that the module has run out of memory to track universes or sources
+     *
+     * If #SACN_DYNAMIC_MEM was defined to 1 when sACN was compiled (the default on non-embedded
+     * platforms), and the configuration you pass to sacn_universe_discovery_create() has source_count_max and
+     * universes_per_source_max set to #SACN_UNIVERSE_DISCOVERY_INFINITE, this callback will never be called and may be
+     * set to NULL.
+     *
+     * If #SACN_DYNAMIC_MEM was defined to 0 when sACN was compiled, source_count_max and universes_per_source_max are
+     * ignored and #SACN_UNIVERSE_DISCOVERY_MAX_SOURCES and #SACN_UNIVERSE_DISCOVERY_MAX_UNIVERSES_PER_SOURCE are used
+     * instead.
+     *
+     * This callback is rate-limited: it will only be called when the first universe discovery packet is received that
+     * takes the module beyond a memory limit.  After that, it will not be called until the number of sources or
+     * universes has dropped below the limit and hits it again.
+     */
+    virtual void HandleMemoryLimitExceeded() = 0;
```

Comment:
This should be optional (not pure virtual). The comment about setting it to NULL is probably copied and pasted from a C callback and should be updated.